boards: Add support for Wio Terminal

joelguittet opened the pull request:
Add support for the Wio Terminal board, including support of its sensors, MicroSD and LCD.

Tested with samples already available and particularly:

- samples/basic/button for the user buttons and led
- samples/sensor/lis2dh for the LIS3DH accelerometer
- samples/subsys/usb/console for the USB (also permit to have a console that the board doesn't provide)
- samples/drivers/display for the LCD
- samples/subsys/fs/fat_fs for the MicroSD Card

The board doesn't support touchscreen/Wifi/BLE yet.
The board doesn't support grove and raspberry pi headers yet (will come shortly).

This Pull Request is following #40811 (not maintained by it's author) which is too much old to be rebased and as a consequence it has been reworked from the very beginnings. There are also much more features such as LCD and MicroSD etc.

joelguittet committed:
boards: Add support for Wio Terminal

Add support for the Wio Terminal board, including support of its
sensors, MicroSD and LCD. Tested with samples already available.

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

kartben commented on boards/arm/wio_terminal/Kconfig.board:
```diff
@@ -0,0 +1,7 @@
+# Wio Terminal board configuration
+
+# SPDX-License-Identifier: Apache-2.0
```
add copyright

kartben commented on boards/arm/wio_terminal/Kconfig.defconfig:
```diff
@@ -0,0 +1,7 @@
+# Wio Terminal  board configuration
+
+# SPDX-License-Identifier: Apache-2.0
```
add copyright

kartben commented on boards/arm/wio_terminal/board.cmake:
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: Apache-2.0
```
add copyright

kartben commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+Hardware
+********
+
+- ATSAMD51P19 ARM Cortex-M4 processor at 120 MHz
```
```suggestion
- ATSAMD51P19 ARM Cortex-M4F processor at 120 MHz (boost up to 200 MHz)
```

kartben commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+- 1x user LED
+- 3x user buttons
+- 5-way user button
+- power/reset/boot mode switch
```
```suggestion
- Power/Reset/Boot mode switch
```

kartben commented on boards/arm/wio_terminal/wio_terminal.dts:
```diff
@@ -0,0 +1,266 @@
+};
+
+&cpu0 {
+	clock-frequency = <120000000>;
```
Since the Wio Terminal handles overclocking like a champ, I would consider changing this to the following for folks to be aware they can tweak this:

```suggestion
	clock-frequency = <120000000>; // 120 MHz (default), can be boosted up to 200 MHz
```

kartben commented on boards/arm/wio_terminal/raspberrypi_40pins_connector.dtsi:
```diff
@@ -0,0 +1,45 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
```
missing copyright header

kartben commented on boards/arm/wio_terminal/grove_connectors.dtsi:
```diff
@@ -0,0 +1,24 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
```
missing copyright header

kartben commented on dts/bindings/gpio/raspberrypi-40pins-header.yaml:
```diff
@@ -0,0 +1,34 @@
+# SPDX-License-Identifier: Apache-2.0
```
missing copyright header

kartben commented on dts/bindings/gpio/raspberrypi-40pins-header.yaml:
```diff
@@ -0,0 +1,34 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    GPIO pins exposing on Raspberry Pi 40 pins headers.
```
```suggestion
    GPIO pins exposed on Raspberry Pi 40 pins headers.
```

nandojve replied:
This should go on a separated PR.

kartben commented on dts/bindings/gpio/grove-header.yaml:
```diff
@@ -0,0 +1,16 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    GPIO pins exposing on Grove 4 pins headers.
```
```suggestion
    GPIO pins exposed on Grove 4-pin headers.
```

nandojve replied:
This should go on a separated PR.

nandojve reviewed (changes requested):
Hi @joelguittet,

Let`s focus on add board first. After that, you can create another two PR one for each new header type + board support.

Please, have a look at [Contribute Guidelines](https://docs.zephyrproject.org/latest/contribute/index.html). This will help to speed up reviews and you have content soon in the tree.

kartben commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+.. image:: img/wio_terminal.png
+     :width: 500px
+     :align: center
+     :alt: Wio Terminal
```
```suggestion
     :alt: Seeed Studio Wio Terminal
```

kartben commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+.. _pinouts:
+    https://wiki.seeedstudio.com/Wio-Terminal-Getting-Started/#pinout-diagram
+
+.. _schematic:
```
```suggestion
.. _schematics:
```

kartben commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+===================
+
+The `Wio Terminal Getting started guide`_ has detailed
+information about the board including `pinouts`_ and the `schematic`_.
```
```suggestion
information about the board including `pinouts`_ and its `schematics`_.
```

kartben commented on boards/arm/wio_terminal/doc/index.rst:
````diff
@@ -0,0 +1,219 @@
+```
+/ {
+    chosen {
+        zephyr,console = &raspberrypi_serial;
````
Why not already put it in wio_terminal.dts by default?

kartben commented:
(apologies for the spam btw, looks like I need to learn how to use the VS Code PR extension better 😨 )

kartben commented on boards/arm/wio_terminal/wio_terminal-pinctrl.dtsi:
```diff
@@ -0,0 +1,82 @@
+
+&pinctrl {
+
+	/* RTL8720D (Wifi/BLE) */
```
I think you want to move these type of comments to the .dts file instead. This here is just to describe the pinmuxes, and not making any assumptions on what peripherals etc are eventually going to be connected.

joelguittet replied:
The schematic do not offer many possibilities. Particularly for Wifi, LCD, SD Card etc. This is why I have added convenient comments here to remind the purpose of the pins.

kartben replied:
I was not suggesting to *not* comment on the purpose of each SERCOM, but rather saying that I think these comments should rather be in the wio_terminal.dts file, as this here is "purely" about how SERCOMs are configured, vs. making any assumption on what they may or may not be attached to. Plus, I would expect most "high-level" users would not even want to dig into wio_terminal-pinctrl.dtsi and would rather be interested in having the wio_terminal.dts file letting them know at a glance what SERCOM corresponds to what peripherals (especially for those peripherals that are not (yet) visible in the tree, such as the RTL chip)

joelguittet replied:
Done

kartben commented on boards/arm/wio_terminal/wio_terminal.dts:
```diff
@@ -0,0 +1,266 @@
+&sercom3 {
+	status = "okay";
+	compatible = "atmel,sam0-i2c";
+	clock-frequency = <I2C_BITRATE_FAST>;
```
should this be `I2C_BITRATE_HIGH`?

joelguittet replied:
There is no strong solution. To be honest I'm in favor of the lowest frequency (100kbits/s), but it's not working due to https://github.com/zephyrproject-rtos/zephyr/issues/42432. So I have set 400kbits/s.
By the way, the user making a real project can configure higher clock if required, making an overlay.
But is it required for a small accelerometer or other Groove sensors ? For most projects I think the answer is no.

kartben commented:
This is awesome, @joelguittet! I feel kind of sad that we ended up doing a lot of work in parallel, apparently (https://github.com/zephyrproject-rtos/zephyr/pull/40811#issuecomment-1433402943) but you definitely went the extra mile with the header definitions etc. 💯 

I've added my comments for you to have a look at, and I agree with @nandojve that the RPi and Grove headers should probably be addressed separately, and through different commits/PRs.

Out of curiosity, have you looked into a driver for the RTL chip? It shouldn't be _too_ hard to generate the eRPC binding and make proper Wi-Fi and Bluetooth drivers, wdyt? https://github.com/Seeed-Studio/seeed-ambd-firmware/tree/master/erpc_idl

joelguittet commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+.. image:: img/wio_terminal.png
+     :width: 500px
+     :align: center
+     :alt: Wio Terminal
```
The name is "Wio Terminal", not "Seed Studio Terminal". I disagree with the suggestion. I propose instead "Seed Studio Wio terminal" if this is suitable.

kartben replied:
Yes sorry that was a typo on my side :) and it's Seeed, not Seed!

joelguittet replied:
So we agree for "Seeed Studio Wio Terminal", correct ?

joelguittet commented on boards/arm/wio_terminal/doc/index.rst:
````diff
@@ -0,0 +1,219 @@
+```
+/ {
+    chosen {
+        zephyr,console = &raspberrypi_serial;
````
Because there is no real console port on the Wio Terminal. This is a serial port available on the Raspberry Pi connector. Not even very easy to connect an FTDI or something similar without a bit of wiring.
So best choice is no console by default. I have added here this comment because many people will look for a console. This is a solution, even if the USB console is very nice too.

joelguittet commented on boards/arm/wio_terminal/wio_terminal.dts:
```diff
@@ -0,0 +1,266 @@
+};
+
+&cpu0 {
+	clock-frequency = <120000000>;
```
I think adding this in the index is enough.

joelguittet replied:
ok, removing

joelguittet commented on boards/arm/wio_terminal/wio_terminal_defconfig:
```diff
@@ -0,0 +1,16 @@
+CONFIG_USE_DT_CODE_PARTITION=y
+CONFIG_GPIO=y
+CONFIG_PINMUX=y
+CONFIG_REGULATOR=y
```
What is this option? I see few board configure this, not sure it'a default choice to be added here?
I don't find clear documentation of the impact. Not described in the other boards for example...

joelguittet replied:
My opinion is also that most people dont' use Windows, so it's not useful. If you are using Windows you can add this config in your own project.

kartben replied:
> What is this option? I see few board configure this, not sure it'a default choice to be added here? I don't find clear documentation of the impact. Not described in the other boards for example...

This allows to get a .uf2 file in the build output directory, which is a convenient way to have a binary that can actually be installed by anyone, by means of putting the Wio Terminal in "boot mode" and copying the file on the mass storage. It is also useful on Windows as the bossa flashing tool is not available (but will soon be, though, see #37538).

kartben replied:
> My opinion is also that most people dont' use Windows, so it's not useful. If you are using Windows you can add this config in your own project.

Except Windows at least 50% of the user base, if not more :)
Even if not enabled by default, it might make a lot of sense to document the ability to turn on this flag in the README, as most people otherwise ignore that they can easily distribute a "flashable" app through this .uf2 format, and I think it is kind of nice for a platform such as Wio Terminal where people might end up sharing the app with people that are not tech-savvy (ex. artists, etc.)

joelguittet replied:
Okay I understand now the usage of the UF2. It's not Windows specific at all! The device mounts as a new disk on any OS when entering bootloader, and it's possible to copy/paste the UF2 file to it to flash the MCU.
And so I agree it's interesting to have it for this kind of boards :-)

However, I'm trying to generate the `zephyr.uf2` file with `CONFIG_BUILD_OUTPUT_UF2=y`, and it's not generated... it depends of `CONFIG_BUILD_OUTPUT_BIN` which is set, so it's not clear for me... Do you know if there is extra requirements ? The `.config` file  still have `# CONFIG_BUILD_OUTPUT_UF2 is not set` even after rebuild all the firmware...

joelguittet replied:
@kartben done

joelguittet commented on boards/arm/wio_terminal/Kconfig.board:
```diff
@@ -0,0 +1,7 @@
+# Wio Terminal board configuration
+
+# SPDX-License-Identifier: Apache-2.0
```
Done

joelguittet commented on boards/arm/wio_terminal/Kconfig.defconfig:
```diff
@@ -0,0 +1,7 @@
+# Wio Terminal  board configuration
+
+# SPDX-License-Identifier: Apache-2.0
```
Done

joelguittet commented on boards/arm/wio_terminal/board.cmake:
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: Apache-2.0
```
Done

joelguittet commented on boards/arm/wio_terminal/grove_connectors.dtsi:
```diff
@@ -0,0 +1,24 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
```
Done

joelguittet commented on boards/arm/wio_terminal/raspberrypi_40pins_connector.dtsi:
```diff
@@ -0,0 +1,45 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
```
Done

joelguittet commented on boards/arm/wio_terminal/wio_terminal-pinctrl.dtsi:
```diff
@@ -0,0 +1,82 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
```
Done

joelguittet commented on dts/bindings/gpio/grove-header.yaml:
```diff
@@ -0,0 +1,16 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    GPIO pins exposing on Grove 4 pins headers.
```
Done

joelguittet commented on dts/bindings/gpio/raspberrypi-40pins-header.yaml:
```diff
@@ -0,0 +1,34 @@
+# SPDX-License-Identifier: Apache-2.0
```
Done

joelguittet replied:
Done

joelguittet commented on dts/bindings/gpio/raspberrypi-40pins-header.yaml:
```diff
@@ -0,0 +1,34 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    GPIO pins exposing on Raspberry Pi 40 pins headers.
```
Done

joelguittet commented:
Hello @nandojve 

I have split the commit into 3 commits onto 3 branches. Thus this PR depends of https://github.com/zephyrproject-rtos/zephyr/pull/55189 which itself depends of https://github.com/zephyrproject-rtos/zephyr/pull/55190. Is it suitable ? Easier to not break the board itself and create 2 PR for the board...

Hello @kartben 

I think I modified or answer for all the reviews you have done, thanks for this exhaustive exercice!
I have read you were having difficulties to work with the SD card, hope my PR helps, do not hesitate to test it :-) I have indicated in the initial message of this PR which example I have used to test the different parts of the board.
Regarding to the Wifi/BLE support, I have checked the details but not work on this part yet. Not sure I will have enough time to do it in the coming weeks. But you are write, this is "basically" adding some libraries etc. And integration with zephyr of course to have the interface like any other network interface. Seed Studio provide c++ libraries, I don't like that, and "Arduino" API which I don't like at all too. Thinking to a pure C integration... We need eRPC, that should be available for other purpose in Zephyr too...

Just one question, I have checked the contributions guidelines, but I'm not sure about the copyright format in the files. I have added something based on other files, but I see all source files in zephyr do not respect the same logic. So i hope this is okay, else tell me I will modify.

Joel

joelguittet commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+The Wio Terminal ships with a the BOSSA compatible UF2 bootloader. The
+bootloader can be entered by quickly tapping the reset button twice.
+
+Additionally, if :code:`CONFIG_USB_CDC_ACM` is enabled then the bootloader
```
Done

joelguittet commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+:zephyr_file:`boards/arm/wio_terminal/wio_terminal_defconfig`.
+
+Zephyr can use the default Cortex-M SYSTICK timer or the SAM0 specific RTC.
+To use the RTC, set :code:`CONFIG_CORTEX_M_SYSTICK=n` and set
```
Done

joelguittet commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+Hardware
+********
+
+- ATSAMD51P19 ARM Cortex-M4 processor at 120 MHz
```
Done

joelguittet commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+- 1x user LED
+- 3x user buttons
+- 5-way user button
+- power/reset/boot mode switch
```
Done

joelguittet commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+===================
+
+The `Wio Terminal Getting started guide`_ has detailed
+information about the board including `pinouts`_ and the `schematic`_.
```
Done

joelguittet commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+
+Zephyr can use the default Cortex-M SYSTICK timer or the SAM0 specific RTC.
+To use the RTC, set :code:`CONFIG_CORTEX_M_SYSTICK=n` and set
+:code:`CONFIG_SYS_CLOCK_TICKS_PER_SEC` to no more than 32 kHZ divided by 7,
```
Done

joelguittet commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,219 @@
+.. _pinouts:
+    https://wiki.seeedstudio.com/Wio-Terminal-Getting-Started/#pinout-diagram
+
+.. _schematic:
```
Done

joelguittet commented:
The pipeline fails on the following in the index:

```
/ {
    chosen {
        zephyr,console = &raspberrypi_serial;
        zephyr,shell-uart = &raspberrypi_serial;
    };
};
```

How to add a piece of code in the index.rst file ?

nandojve commented:
> Hello @nandojve
> 
> I have split the commit into 3 commits onto 3 branches. Thus this PR depends of #55189 which itself depends of #55190. Is it suitable ? Easier to not break the board itself and create 2 PR for the board...

Hi @joelguittet ,

In my opinion this PR should not depends on #55189 and #55190. The connectors should not be mandatory to you start with this board. I was looking at seeed website and I think best is use CDC-ACM as default console too. With that, users can start using this board with only a USB cable, if I understood correctly. If that is the case, it should be the easy way to support it in the project. Remember that users always have the ability to add overlays to change default devicetree properties,

nandojve commented:
> > [...] and I think best is use CDC-ACM as default console too. With that, users can start using this board with only a USB cable, if I understood correctly. If that is the case, it should be the easy way to support it in the project.
> 
> +1, this is definitely how I'm using it and I think it should be the default. I have to admit I'm actually not sure what you meant when you originally said "also permit to have a console that the board doesn't provide", @joelguittet?

Maybe this could help https://github.com/zephyrproject-rtos/zephyr/blob/f77f02c03259988c3b7289f11dd7d831a1ab7ce8/boards/arm/bl654_usb/bl654_usb.dts#L18-L19

joelguittet commented:
Hello @nandojve and @kartben 

 "also permit to have a console that the board doesn't provide" I mean there is no dedicated serial port with an FTDI or something similar to get logs. The user have to use an UART on the RPI connector or the USB link.

Regarding to the console:
- it's fine for me to have it on USB by default
- but is it possible to use USB for something else in any application then ? I mean if the user wants USB for networking, mass storage, hid, etc, is it still possible ?

Regarding to the headers:
- @nandojve I don't understand, is it a contribution guidelines issue ? Or your advice is to not have connectors at all ?
- Plenty of STM32 NUCELO boards have the arduino_r3 header definition, so the usage should be the same here. It simplify greatly the usage of I2C/SPI/UART/GPIOs when using shields or hats for example.

Thanks
Joel

nandojve commented:
> but is it possible to use USB for something else in any application then ? I mean if the user wants USB for networking, mass storage, hid, etc, is it still possible ?

I think it won't be a problem since Zephyr supports USB Composite.

> I don't understand, is it a contribution guidelines issue ?

It is a good practice send small PRs that touch less code to involve less possible maintainers. In this specific case, you are introducing the board + 2 new headers and you should, preferable, do one each time. This is why I mentioned that you should focus first in add the board and than later introduce the headers + board dts entries. Usually different maintainers may have different needs and could take long review process.

> Or your advice is to not have connectors at all ?

My advice is add connectors in another moment to keep this simpler.

joelguittet commented:
@nandojve PR updated, do not depend anymore of the raspberrypi and grove headers actually.
Regarding to the console over USB, I have checked your reference and also other. I disagree modifying the dts and defconfig is enough. There are much more code required to get the console working (I have just tested). See: https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/usb/console and particularly: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/usb/console/src/main.c
So I think the BL654 board is not a good example. Maybe it's old and outdated or it's simply not working out of the box.

nandojve commented:
>  PR updated, do not depend anymore of the raspberrypi and grove headers actually.

I think you forgot to remove *_connector.dtsi files.

> Regarding to the console over USB, I have checked your reference and also other. I disagree modifying the dts and defconfig is enough. There are much more code required to get the console working (I have just tested).

This is not a block from my side, it is just a suggestion. But I agreed that currently USB console could be strange. You can check below snip. It is enough to get USB console to work. The details is that **samples/hello_world** is not a good application to demonstrate functionality in case of USB console. If you decide go with this, I'll suggest select another example like **samples/philosophers** in the documentation. The USB console seems to be used without used in this case without handshake and application starts ASAP and sent data to host, which will be probably lost since host takes a few time to enumerate and make the port available. The only thing that rely matters is that users should understand those details to create their own application correctly since there is nothing wrong.

```diff
(.venv-3.10) gfbudke@debian12:~/zephyr/mainline/zephyr$ git diff
diff --git a/boards/arm/atsamr21_xpro/atsamr21_xpro.dts b/boards/arm/atsamr21_xpro/atsamr21_xpro.dts
index 29d942aa79..935da7e898 100644
--- a/boards/arm/atsamr21_xpro/atsamr21_xpro.dts
+++ b/boards/arm/atsamr21_xpro/atsamr21_xpro.dts
@@ -13,8 +13,8 @@
        compatible = "atsamr21,xpro", "atmel,samr21g18a", "atmel,samr21";
 
        chosen {
-               zephyr,console = &sercom0;
-               zephyr,shell-uart = &sercom0;
+               zephyr,console = &cdc_acm_uart;
+               zephyr,shell-uart = &cdc_acm_uart;
                zephyr,sram = &sram0;
                zephyr,flash = &flash0;
                zephyr,ieee802154 = &ieee802154;
@@ -188,6 +188,10 @@ zephyr_udc0: &usb0 {
 
        pinctrl-0 = <&usb_dc_default>;
        pinctrl-names = "default";
+
+       cdc_acm_uart: cdc_acm_uart {
+               compatible = "zephyr,cdc-acm-uart";
+       };
 };
 
 ext1_spi: &sercom5 {
diff --git a/boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig b/boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig
index b327352ddc..341b460555 100644
--- a/boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig
+++ b/boards/arm/atsamr21_xpro/atsamr21_xpro_defconfig
@@ -12,3 +12,6 @@ CONFIG_UART_INTERRUPT_DRIVEN=y
 CONFIG_GPIO=y
 CONFIG_SOC_ATMEL_SAMD_XOSC32K=y
 CONFIG_SOC_ATMEL_SAMD_XOSC32K_AS_MAIN=y
+
+CONFIG_USB_DEVICE_STACK=y
+CONFIG_USB_DEVICE_INITIALIZE_AT_BOOT=y
```

To get a better shell it may good add:
```diff
+CONFIG_UART_LINE_CTRL=y
+CONFIG_SHELL=y
+CONFIG_SHELL_BACKEND_SERIAL_CHECK_DTR=y
```

joelguittet commented:
> I think you forgot to remove *_connector.dtsi files.

Correct, I have not pushed yet, trying to get the USB console working so I have pending modification in the file.

> You can check below snip. It is enough to get USB console to work

No, still not working. No serial port enumerated at all on host PC.

joelguittet commented:
@kartben @nandojve 

Okay, trying to activate UF2 and USB console from the board config. The options I set to `y` in the `wio_terminal_defconfig` file remains `is not set` in the `build/zephyr/.config`. Do you know a possible reason ?

For example in the defconfig:
```
CONFIG_BUILD_OUTPUT_UF2=y
```

And in the `.config`:
```
# CONFIG_BUILD_OUTPUT_UF2 is not set
```

I have updated my fork on top of main, I have executed `west update` and deleted the `build` directory to build from nothing.

nandojve commented:
You need enable console to USB console work:

```console
(.venv-3.10) gfbudke@debian12:~/zephyr/mainline/zephyr$ git lg -1
* 9201c07b1b - (HEAD -> 54953) boards: Add support for Wio Terminal (2 days ago) <Joel Guittet>

```
```diff
(.venv-3.10) gfbudke@debian12:~/zephyr/mainline/zephyr$ git diff
diff --git a/boards/arm/wio_terminal/wio_terminal.dts b/boards/arm/wio_terminal/wio_terminal.dts
index 0806a2d10a..2a16a296ab 100644
--- a/boards/arm/wio_terminal/wio_terminal.dts
+++ b/boards/arm/wio_terminal/wio_terminal.dts
@@ -19,6 +19,8 @@
                zephyr,flash = &flash0;
                zephyr,code-partition = &code_partition;
                zephyr,display = &ili9341;
+               zephyr,console = &wio_cdc_acm_console;
+               zephyr,shell-uart = &wio_cdc_acm_console;
        };
 
        /* These aliases are provided for compatibility with samples */
@@ -264,4 +266,8 @@ zephyr_udc0: &usb0 {
        status = "okay";
        pinctrl-0 = <&usb_dc_default>;
        pinctrl-names = "default";
+
+       wio_cdc_acm_console: wio_cdc_acm_console {
+               compatible = "zephyr,cdc-acm-uart";
+       };
 };
diff --git a/boards/arm/wio_terminal/wio_terminal_defconfig b/boards/arm/wio_terminal/wio_terminal_defconfig
index bfc84a135e..7827a2f3be 100644
--- a/boards/arm/wio_terminal/wio_terminal_defconfig
+++ b/boards/arm/wio_terminal/wio_terminal_defconfig
@@ -14,3 +14,11 @@ CONFIG_BOOTLOADER_BOSSA_ADAFRUIT_UF2=y
 CONFIG_BUILD_OUTPUT_UF2=y
 CONFIG_GPIO=y
 CONFIG_REGULATOR=y
+
+# Console USB CDC-ACM
+CONFIG_SERIAL=y
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_USB_DEVICE_STACK=y
+CONFIG_USB_DEVICE_INITIALIZE_AT_BOOT=y
```

BTW, UF2 works for me!

```console
...
-- Zephyr version: 3.3.99 (/home/gfbudke/zephyr/mainline/zephyr), build: zephyr-v3.3.0-672-g9201c07b1bb6
[150/160] Linking C executable zephyr/zephyr_pre0.elf

[154/160] Linking C executable zephyr/zephyr_pre1.elf

[160/160] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       27092 B       496 KB      5.33%
             RAM:       11136 B       192 KB      5.66%
        IDT_LIST:          0 GB         2 KB      0.00%
Converting to uf2, output size: 54272, start address: 0x4000
Wrote 54272 bytes to zephyr.uf2
```

joelguittet commented:
@nandojve ok thanks, and you are ... lucky.
I have the same as you and I have no USB CDC console, and no UF2 file.
I just pushed my branch with all my modifications so you can review the PR and please can you confirm me you get the UF2 and the USB CDC console properly configured in the `.config` file ?
Can you also indicate me with sample you build to get this uf2 file ? maybe it's not working for all samples ?

joelguittet commented:
@nandojve found my issue: I was working in a separated directory `/zephyrproject/zephyr_joelguittet`, it was making a mix or something with `/zephyrproject/zephyr`, dispite the fact I was building from the first one.
UF2 generated and USB CDC Console working on my side.

nandojve commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,199 @@
+The wio_terminal board configuration supports the following
+hardware features:
+
++-----------+------------+------------------------------------------+
+| Interface | Controller | Driver/Component                         |
++===========+============+==========================================+
+| NVIC      | on-chip    | Nested vector interrupt controller       |
++-----------+------------+------------------------------------------+
+| Flash     | on-chip    | Can be used with LittleFS to store files |
++-----------+------------+------------------------------------------+
+| SYSTICK   | on-chip    | Systick                                  |
++-----------+------------+------------------------------------------+
+| WDT       | on-chip    | Watchdog                                 |
++-----------+------------+------------------------------------------+
+| GPIO      | on-chip    | I/O ports                                |
++-----------+------------+------------------------------------------+
+| USART     | on-chip    | Serial port                              |
++-----------+------------+------------------------------------------+
+| I2C       | on-chip    | Inter-Integrated Circuit                 |
++-----------+------------+------------------------------------------+
+| SPI       | on-chip    | Serial Peripheral Interface port         |
++-----------+------------+------------------------------------------+
+| TRNG      | on-chip    | True Random Number Generator             |
++-----------+------------+------------------------------------------+
+| HWINFO    | on-chip    | Unique 128 bit serial number             |
++-----------+------------+------------------------------------------+
+| RTC       | on-chip    | Real-Time Counter                        |
++-----------+------------+------------------------------------------+
+| USB       | on-chip    | USB device                               |
++-----------+------------+------------------------------------------+
+| PWM       | on-chip    | PWM                                      |
++-----------+------------+------------------------------------------+
```
Update table to new format, see [atsamr34_xpro](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/atsamr34_xpro/doc/index.rst) as an example.

joelguittet replied:
Done

nandojve commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,199 @@
+The SAMD51 MCU is configured to use the 32.768 kHz internal oscillator
+with the on-chip PLL generating the 120 MHz system clock.
+
+Serial Port
+===========
+
+The SAMD51 MCU has 8 SERCOM based USARTs. On the Wio Terminal, no SERCOM is
+available for the Zephyr console and indeed it is configured by default over
+the USB interface.
```
Console will be by USB, require update this section.

joelguittet replied:
Done

nandojve commented on boards/arm/wio_terminal/wio_terminal.dts:
```diff
@@ -0,0 +1,271 @@
+	pinctrl-names = "default";
+	cs-gpios = <&portc 19 GPIO_ACTIVE_LOW>;
+
+  /* MicroSD Card */
+	sdhc0: sdhc@0 {
```
comment Identation...

joelguittet replied:
Done

nandojve commented on boards/arm/wio_terminal/wio_terminal.dts:
```diff
@@ -0,0 +1,271 @@
+	pinctrl-0 = <&usb_dc_default>;
+	pinctrl-names = "default";
+
+	cdc_acm_uart0: cdc_acm_uart0 {
+		compatible = "zephyr,cdc-acm-uart";
+	};
```
Do not use generic names here. It is better use a prefix like `wio_terminal_console`

joelguittet replied:
Done

nandojve commented on boards/arm/wio_terminal/wio_terminal_defconfig:
```diff
@@ -0,0 +1,29 @@
+CONFIG_BOOTLOADER_BOSSA=y
+CONFIG_BOOTLOADER_BOSSA_ADAFRUIT_UF2=y
+CONFIG_BUILD_OUTPUT_UF2=y
+CONFIG_REGULATOR=y
```
`CONFIG_REGULATOR=y` Why do you need regulators here?

joelguittet replied:
Some power supplies are driven by the MCU: raspberry pi header power supply, lcd backlight, usb host

nandojve commented on boards/arm/wio_terminal/wio_terminal.yaml:
```diff
@@ -0,0 +1,10 @@
+toolchain:
+  - zephyr
+  - gnuarmemb
+  - xtools
```
I think still necessary add HW supported features, right?

https://github.com/zephyrproject-rtos/zephyr/blob/a2541d6c82437b02fb26db5e8dd1928e5fd0b1d5/boards/arm/atsaml21_xpro/atsaml21_xpro.yaml#L13-L23

joelguittet replied:
yes! will add now

nandojve commented on boards/arm/wio_terminal/wio_terminal_defconfig:
```diff
@@ -0,0 +1,29 @@
+CONFIG_ARM_MPU=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_HW_STACK_PROTECTION=y
+CONFIG_BOOTLOADER_BOSSA=y
+CONFIG_BOOTLOADER_BOSSA_ADAFRUIT_UF2=y
+CONFIG_BUILD_OUTPUT_UF2=y
```
Move BOSSA config to a specific section like Console. This could be good for future references.

nandojve reviewed (dismissed):
Hi @joelguittet ,

LGTM!

kartben commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,212 @@
+=========
+
+In addition to the built-in bootloader, the Wio Terminal can be flashed and
+debugged using a SWD probe such as the Segger J-Link.
```
```suggestion
debugged using an SWD probe such as the Segger J-Link.
```

kartben commented on boards/arm/wio_terminal/wio_terminal.dts:
```diff
@@ -0,0 +1,271 @@
+	status = "okay";
+};
+
+&flash0 {
```
Can you confirm that it's not possible to hook up the W25Q32JVZPIM external flash at the moment? Or is it? It's using QSPI, so I'm not sure that's supported in Zephyr for SAM0? Appreciate your feedback here as it would be super cool if it can be added.

joelguittet replied:
I have checked for this and this is exactly the point: there is no QSPI driver for sam0 actually in Zephyr. This need to be developed. I have understood the atmel parts of Zephyr is mainly developed/maintained by contributors.

nandojve replied:
Hi Folks,

Independently if feature is available or nor (in this case not) it is important first bring board and later add feature by feature. This way we involve only people responsible for the area which speed up acceptance.

Just to mention, see #39991

kartben replied:
Yes, 100% agreed :)

kartben commented on boards/arm/wio_terminal/wio_terminal_defconfig:
```diff
@@ -0,0 +1,31 @@
+CONFIG_USB_DEVICE_STACK=y
+CONFIG_USB_DEVICE_INITIALIZE_AT_BOOT=y
+CONFIG_USB_DEVICE_VID=0x2886
+CONFIG_USB_DEVICE_PID=0x002d
```
```suggestion
CONFIG_USB_DEVICE_PID=0x802D
```
as per https://github.com/Seeed-Studio/ArduinoCore-samd/blob/master/boards.txt#L53

joelguittet replied:
Good point, integrating

kartben commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,212 @@
+
+The Wio Terminal is a small (72 mm x 57 mm x 12 mm) and powerful ARM board with
+wireless connectivity (2.4G/5G Dual Bands Wireless and BLE5.0), LCD display,
+USB C port, FPC connector, MicroSD card slot, Raspberry pi compatible 40pin
```
```suggestion
USB-C port, FPC connector, microSD card slot, Raspberry Pi compatible 40-pin
```

kartben commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,212 @@
+********
+
+The Wio Terminal is a small (72 mm x 57 mm x 12 mm) and powerful ARM board with
+wireless connectivity (2.4G/5G Dual Bands Wireless and BLE5.0), LCD display,
```
```suggestion
wireless connectivity (2.4G/5G dual-band Wi-Fi and BLE 5.0), LCD display,
```

kartben reviewed (changes requested):
This is awesome work Joel! A final round of comments/edits on my side.
I'm really excited about this one, as this will make for an awesome board for showcasing GUI stuff! (even more so if we can get the 4MB external flash to work too -- see my comment on that topic.

kartben commented on boards/arm/wio_terminal/doc/index.rst:
```diff
@@ -0,0 +1,212 @@
+The Wio Terminal ships with a the BOSSA compatible UF2 bootloader. The
+bootloader can be entered by quickly tapping the reset button twice.
+
+The UF2 file is generated when building, it is possible to use it to flash the
```
@joelguittet sorry, look like I should have suggested the change differently, as it was more than one line. Could you please incorporate the full change, as it fixes improper grammar issues in the rest of the paragraph?

kartben commented on boards/arm/wio_terminal/wio_terminal.dts:
```diff
@@ -0,0 +1,280 @@
+		reset-gpios = <&portc 7 GPIO_ACTIVE_LOW>;
+		pixel-format = <ILI9XXX_PIXEL_FORMAT_RGB565>;
+		rotation = <270>;
+		width = <240>;
```
I've just tested some GUI code and realized you will want to swap width and height here

joelguittet replied:
Good catch! I have just tested with `samples/drivers/display` and it makes no differences.
Are you testing with a sample from Zephyr or something else ?
Will do the modification right now.

kartben replied:
The issue occurs with samples/subsys/display/lvgl (which I guess is actually taking the width/height information into account vs. the "dumb" driver sample) -- it shows a white screen when the width and height are wrong (in fact I think the app even crashes but I am too lazy to test again).

joelguittet replied:
Checked with this sample, yeah I agree. Modification has been done, sample is properly displaying the hello world :-)

joelguittet commented:
@kartben hello, it seems all comments are integrated, please help with your review (the PR currently displays you requested changes :) )
Thanks,
Joel

joelguittet committed:
doc: release_notes v3.3: Add new board wio terminal

Add new board Wio Terminal from Seeed Studio

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

kartben commented:
> @kartben hello, it seems all comments are integrated, please help with your review (the PR currently displays you requested changes :) ) Thanks, Joel

There is this one, which I think is the only thing left on my side :) https://github.com/zephyrproject-rtos/zephyr/pull/54953#discussion_r1126656220 - I am guessing the conversation got muted after you had marked the comment as resolved? (btw it's usually better to let the original review resolve the comments).

I would suggest to use the opportunity of this one last change to update the future release notes with a mention of your awesome work :)

https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-3.4.rst

```
* Made these changes for ARM boards:

  * Seeed Studio Wio Terminal
```

joelguittet commented:
@kartben done :-)

nandojve commented:
CI must be green

kartben commented:
I am excited to see this getting merged! 
@nashif @carlescufi fwiw I've tested this extensively over the past few weeks so let us know when you have a chance to +1

joelguittet commented:
Rebased, ready to merge :-) Thanks at @nandojve and @kartben for the detailed reviews!
